### DISCUSSION TITLE
Group records using lookup object reference fields in Domain?

### DISCUSSION THREAD

**Discussion by pp1-logitech** [apex-enterprise-patterns/fflib-apex-common#448] [Q&A]
I have a use case to group the accounts with the DNB Company Records(lookup in account)'s Global_Ultimate_DUNS_Number__c and check if the accounts tied via Global DUNS if they are present in more than one country and employee count > 10000 set the Segmentation to "Global" else set it to "Enterprise".

Here's what i tried:
`

        IAccounts accountsWithGlobalDuns =  Accounts.newInstance(accountsWithGlobalDuns);

        Set<Id> dnbIds = globalDUNSAccounts.getDNBIds();

       // gives the dnbrecord id to global duns where dnbid is the id tied to the global ultimate account
        Map<Id, String> gdnbIdToGduns = DnBCompanyRecords.newInstance(
            DnBCompanyRecordsSelector.newInstance().selectById(dnbIds)
        ).getDNBIdByGDUNSWithDUNSSameAsGDUNS(); 

        //gives the dnbrecordid to its global duns association
        Map<Id, String> dnbIdToGduns = DnBCompanyRecords.newInstance(
            DnBCompanyRecordsSelector.newInstance().selectById(dnbIds)
        ).getDNBIdByGDUNS();

       //filters only the globalultimateaccounts
        IAccounts globalUtimateAccounts =  accountsWithGlobalDuns.selectByDNBId(gdnbIdToGduns.keySet());       
        
        Map<String, Decimal> gdunsByCountryCount = accountsWithGlobalDuns.getCountryCountByGDUNS(dnbIdToGduns)

        IAccounts globalUtimateAccountsInMoreThanOneCountry =  globalUtimateAccounts .selectByDnbIdAndGDUNSMoreThanCountryCount(
         1, dnbIdToGduns, gdunsByCountryCount 
        ).selectByEmployeesGreaterThan(10000).setSegmentation('Global');

`
Accounts:

`
              
             selectByDnbIdAndGDUNSMoreThanCountryCount(Decimal countryCount, Map<Id, String> idToGDUNS, Map<String, Decimal> 
       gdundsToCount) {

                for (Account account : getAccounts){

                    if (gdundsToCount.get(idToGDUNS.get(account.Dnb_Company_Record__c)) > countryCount) {
                   // add 

                }

                return new Accounts(<list>);

        }
`

I feel like above is very complex. Is there any proper way to do this in fflib? So, can we use lookup field's reference in another domain class?

**Comment by ImJohnMDaniel** [comment]
G'day @pp1-logitech, it is difficult to say if there is a more streamlined approach with out a look at more of the code.  Having said that, there are at least suggestions that I would make:
* You don't need to initialize the domains with specifically executing the selector.  Simply supplying the Set<Id> to the domain newInstance method should suffice.
* Reuse the same domain since they are the same records
